contracts/profiles/StealthKeyResolver.sol

```diff
@@ -4,36 +4,64 @@ import "../ResolverBase.sol";
 
-    struct StealthKey {
-        uint256 generationPubKey;
-        uint256 encryptionPubKey;
-    }
-
-    mapping(bytes32=>StealthKey) _stealthKeys;
+    // node => prefix => key
+    mapping(bytes32 => mapping(uint256 => uint256)) _stealthKeys;
 
     /**
      * Sets the stealth keys associated with an ENS name, for anonymous sends.
      * May only be called by the owner of that node in the ENS registry.
      * @param node The node to update.
-     * @param generationPubKey The public key for generating a stealth address
-     * @param encryptionPubKey The public key to use for encryption
+     * @param spendingPubKeyPrefix Prefix of the spending public key (2 or 3)
+     * @param spendingPubKey The public key for generating a stealth address
+     * @param viewingPubKeyPrefix Prefix of the viewing public key (2 or 3)
+     * @param viewingPubKey The public key to use for encryption
      */
-    function setStealthKeys(bytes32 node, uint256 generationPubKey, uint256 encryptionPubKey) external authorised(node) {
-        _stealthKeys[node] = StealthKey(generationPubKey, encryptionPubKey);
-        emit StealthKeyChanged(node, generationPubKey, encryptionPubKey);
+    function setStealthKeys(bytes32 node, uint256 spendingPubKeyPrefix, uint256 spendingPubKey, uint256 viewingPubKeyPrefix, uint256 viewingPubKey) external authorised(node) {
+        require(
+            (spendingPubKeyPrefix == 2 || spendingPubKeyPrefix == 3) &&
+            (viewingPubKeyPrefix == 2 || viewingPubKeyPrefix == 3),
+            "StealthKeyResolver: Invalid Prefix"
+        );
+
+        delete _stealthKeys[node][0];
+        delete _stealthKeys[node][1];
+        delete _stealthKeys[node][2];
+        delete _stealthKeys[node][3];
+
+        _stealthKeys[node][spendingPubKeyPrefix - 2] = spendingPubKey;
+        _stealthKeys[node][viewingPubKeyPrefix] = viewingPubKey;
+
+        emit StealthKeyChanged(node, spendingPubKeyPrefix, spendingPubKey, viewingPubKeyPrefix, viewingPubKey);
     }
 
     /**
      * Returns the stealth key associated with a name.
      * @param node The ENS node to query.
-     * @return generationPubKey The public key for generating a stealth address.
-     * @return encryptionPubKey The public key to use for encryption.
+     * @return spendingPubKeyPrefix Prefix of the spending public key (2 or 3)
+     * @return spendingPubKey The public key for generating a stealth address
+     * @return viewingPubKeyPrefix Prefix of the viewing public key (2 or 3)
+     * @return viewingPubKey The public key to use for encryption
      */
-    function stealthKeys(bytes32 node) external view returns (uint256 generationPubKey, uint256 encryptionPubKey) {
-        StealthKey memory key = _stealthKeys[node];
-        return (key.generationPubKey, key.encryptionPubKey);
+    function stealthKeys(bytes32 node) external view returns (uint256 spendingPubKeyPrefix, uint256 spendingPubKey, uint256 viewingPubKeyPrefix, uint256 viewingPubKey) {
```

Comment:
Would it make sense to have these functions just set and get a 4-element array, and leave the semantics up to the caller?

Reply:
My first reaction is that it's preferable to codify the semantics, rather than leaving them up to the caller, to remove any ambiguity or sources of error between apps. But I'd be interested in hearing your thoughts on why leaving them up to the caller might be preferable.

Reply:
hey @Arachnid friendly bump on this :)